src/vm/checker/typecheck/mod.rs

```diff
@@ -0,0 +1,404 @@
+        self.type_map.set_type(&map_expression[1], no_type())?;
+        self.type_map.set_type(&map_expression[2], no_type())?;
+        self.type_map.set_type(&map_expression[3], no_type())?;
+        // should we set the type of the subexpressions of the signature to no-type as well?
+
+        let map_name = map_expression[1].match_atom()
+            .ok_or(CheckError::new(CheckErrors::BadMapName))?;
+        let key_type = &map_expression[2];
+        let value_type = &map_expression[3];
+
+        let key_type = TypeSignature::new_tuple(
+            TupleTypeSignature::parse_name_type_pair_list(key_type)
+                .map_err(|_| { CheckError::new(CheckErrors::BadMapTypeDefinition) })?)
+            .map_err(|_| { CheckError::new(CheckErrors::BadMapTypeDefinition) })?;
+        let value_type = TypeSignature::new_tuple(
+            TupleTypeSignature::parse_name_type_pair_list(value_type)
+                .map_err(|_| { CheckError::new(CheckErrors::BadMapTypeDefinition) })?)
+            .map_err(|_| { CheckError::new(CheckErrors::BadMapTypeDefinition) })?;
+
+
+        Ok((map_name.to_string(), (key_type, value_type)))
+    }
+
+    // Aaron: note, using lazy statics here would speed things up a bit and reduce clone()s
+    fn try_native_function_check(&mut self, function: &str, args: &[SymbolicExpression], context: &TypingContext) -> Option<TypeResult> {
+        if let Some(ref native_function) = NativeFunctions::lookup_by_name(function) {
+            let typed_function = TypedNativeFunction::type_native_function(native_function);
+            Some(typed_function.type_check_appliction(self, args, context))
+        } else {
+            None
+        }
+    }
+
+    fn type_check_function_application(&mut self, expression: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+        if let Some((function_name, args)) = expression.split_first() {
+            self.type_map.set_type(function_name, no_type())?;
+            let function_name = function_name.match_atom()
+                .ok_or(CheckError::new(CheckErrors::NonFunctionApplication))?;
+
+            if let Some(type_result) = self.try_native_function_check(function_name, args, context) {
+                type_result
+            } else {
+                let function_type = self.get_function_type(function_name)
+                    .ok_or(CheckError::new(CheckErrors::UnknownFunction(function_name.clone())))?;
+                self.type_check_function_type(&function_type, args, context)
+            }
+        } else {
+            Err(CheckError::new(CheckErrors::NonFunctionApplication))
+        }
+    }
+
+    pub fn type_check(&mut self, expr: &SymbolicExpression, context: &TypingContext) -> TypeResult {
+        let mut result = self.inner_type_check(expr, context);
+
+        if let Err(ref mut error) = result {
+            if !error.has_expression() {
+                error.set_expression(expr);
```

Comment:
I'm confused by what's happening here.  Is this how stack traces are built?

Reply:
Not stack traces, but a pointer to the offending subexpression -- this will allow us to either (1) print out the offending subexpression, or (2) if the parser includes line numbers, print line numbers as well.